Enhance docs with GitBook (#398)
src-d/lookout: docs/how-to-run.md

# Running lookout

The recommended way to locally try **source{d} lookout** is using `docker-compose` as described in [quickstart documentation](../README.md#quickstart).

> In some situations, you might want to run all its dependencies and components separately, or you can even want to [run it in a distributed mode](#running-lookout-in-distributed-mode). Both ways to run lookout would require to follow these general steps:

Reviewer: ```suggestion
In some situations, you might want to run all its dependencies and components separately, or you may even want to [run it in a distributed mode](#running-lookout-in-distributed-mode). Both ways to run lookout would require to follow these general steps:
```

> 1. [Run lookout dependencies](#dependencies),

Reviewer 2: similar to the comments before, let's drop `,`

1. [Run the desired analyzers](#run-the-analyzers),
1. [Configure lookout](#configure-lookout),
1. [Run lookout](#run-lookout).

> In other situations you just want to locally test an analyzer without accessing GitHub at all; for that situations, you might want to read [`lookout-sdk` binary documentation](lookout-sdk.md).

Reviewer: ```suggestion
In other situations you may just want to test an analyzer locally without accessing GitHub at all. For those situations, you might want to read the [`lookout-sdk` binary documentation](lookout-sdk.md).
```

## Running lookout in Distributed Mode

lookout can be also run in a distributed fashion using a [RabbitMQ](https://www.rabbitmq.com/) queue to coordinate a _watcher_ and several _workers_.

- The _watcher_ process monitors GitHub pull requests and enqueues jobs for new events.

> - The running _workers_ dequeue jobs as they become available calling the registered analyzers and posting its returned comments.

Reviewer: ```suggestion
- The running _workers_ dequeue jobs as they become available, call the registered analyzers, and post the results as comments.
```

The general steps to run source{d} lookout in distributed mode are the same as said above.

# Dependencies

**lookout** needs a running instance of:

* [bblfshd](https://github.com/bblfsh/bblfshd) to parse files into [UAST](https://doc.bblf.sh/uast/uast-specification.html).

> * [PostgreSQL](https://www.postgresql.org) to internally store `PushEvent`, `ReviewEvent`, `ReviewTarget` and `Comment`.

Reviewer: The list of things seems out of place. The DB is managed by lookout, as a user you don't need to know or care about `PushEvent`, `ReviewEvent`, `ReviewTarget` and `Comment`. Adding them here makes it look like it's relevant for you to know about these things, but we don't explain what is each one (because you don't really need to know about them).

In this case less is more in my opinion.
```suggestion
* [PostgreSQL](https://www.postgresql.org) for persistence.
```

* (**optional**) [RabbitMQ](https://www.rabbitmq.com) to coordinate a _watcher_ and several _workers_ (when running lookout in a distributed way).

You can run them manually or with `docker-compose`.

```bash
$ docker-compose up -d --force-recreate bblfsh postgres
```

In case you want to run it in a distributed way, you will also need RabbitMQ, so you can run instead:

```bash

> $ docker-compose -f docker-compose.yaml -f docker-compose-rabbitmq.yml up -d --force-recreate bblfsh postgres rabbitmq

Reviewer: Do we really want to provide a docker-compose file with rabbitmq? I see its value for us, as developers. But for someone looking to deploy lookout in a distributed way, will they care about running the distributed scenario in their own laptop?

On the other hand taking a look at the compose file can be helpful to understand how to deploy each component manually. :thinking:

Author: I think it worth having it in a separated file, that can be used if needed, and is consistent with how different dependencies are run.

1. add the URLs of the repositories to be watched or authenticate as a GitHub App.

# Run lookout

Download the latest `lookoutd` binary from [lookout releases page](https://github.com/src-d/lookout/releases).

## Initialize the Database

_For non-default configuration, please take a look into [**`lookoutd` Command Options**](#options)_

```bash
$ lookoutd migrate
```

## Start lookout

_For non-default configuration, please take a look into [**`lookoutd` Command Options**](#options)_

For a single server watching GitHub and processing events, just run:

```bash
$ lookoutd serve [--dry-run] [--github-token=<token> --github-user=<user>]
```

### Distributed Mode

_For non-default configuration, please take a look into [**`lookoutd` Command Options**](#options)_

In order to run it in a distributed mode, the _watcher_ and the _workers_ must be run separately.

Run the _watcher_:

```bash
$ lookoutd watch [--github-token=<token> --github-user=<user>]
```

and as many _workers_ you need:

```bash
$ lookoutd work [--dry-run] [--github-token=<token> --github-user=<user>]
```

<a id=options></a>
# Appendix: `lookoutd` Command Options

> `lookoutd` binary include some subcommands as described above, and they accept many different options; you can use:

Reviewer: ```suggestion
`lookoutd` binary includes some subcommands as described above, and they accept many different options; you can use:
```

- `lookoutd -h`, to see all the available subcommands.
- `lookoutd subcommand -h`, to see all the options for the given subcommand.

Here are some of the most relevant options for `lookoutd`:

> - [dry-run mode](#dry-run-mode),

Reviewer: I'd remove the trailing `,`

- [authentication options](#authentication-options),
- [number of concurrent events to process](#concurrent-events),
- [dependencies URIs](#dependencies-uris)
- [logging options](#logging-options)

## Dry-run Mode
If you want to avoid posting the analysis results on GitHub, and only print them, enable the _dry-run_ mode when running `serve`, `work` subcommands:

> | subcommands | Env var | Option |

Reviewer 2: looks like tables are broken:
https://github.com/src-d/lookout/blob/fdd5d732139120ce7e87f2e370de9ec6c55fbd0f/docs/how-to-run.md

Author: whaaat? how did I ruin them!!?
thanks for spotting that problem!

Author: fixed by 3912bed

Author: fixed by e3d47eb
